File: src/RESTworld/RESTworld.Client.Angular/projects/ngx-restworld-client/src/lib/services/restworld-client.ts
import { HttpHeaders, HttpResponse } from '@angular/common/http';
import * as _ from "lodash";
import { FormsResource, HalClient, Link, PagedListResource, Resource, ResourceDto, Template } from "@wertzui/ngx-hal-client";
import { LinkNames } from "../constants/link-names";
import { ProblemDetails } from "../models/problem-details";
import { RESTworldOptions } from "../models/restworld-options";

export class RESTworldClient {

  private _homeResource?: Resource;
  private _defaultCurie?: string;
  public get halClient() {
    return this._halClient;
  }

  constructor(
    private _halClient: HalClient,
    private _options: RESTworldOptions
  ) { }

  public async ensureHomeResourceIsSet(): Promise<void> {
    if (!this._homeResource) {
      const response = await this.getHomeForced();
      if (ProblemDetails.isProblemDetails(response.body)) {
        throw new Error(`Cannot get the home resource from ${this._options.BaseUrl} with Version ${this._options.Version}. Response was: ${response}`);
      }
      if (!response.body)
        throw new Error(`Cannot get the home resource from ${this._options.BaseUrl} with Version ${this._options.Version}. Response was empty.`);
      this._homeResource = response.body;
      this.setDefaultCurie();
    }
  }

  private async getHomeForced(): Promise<HttpResponse<Resource | ProblemDetails>> {
    const header = RESTworldClient.createHeaders('application/hal+json', this._options.Version);
    const response = await this.halClient.get(this._options.BaseUrl, Resource, ProblemDetails, header);
    return response;
  }

  private setDefaultCurie(): void {
    const curies = this._homeResource?._links?.curies;
    if (!curies || curies.length === 0 || !curies[0])
      this._defaultCurie = undefined;
    else
      this._defaultCurie = curies[0].name;
  }

  public async getList<TListDto extends ResourceDto>(rel: string, parameters: {}, headers?: HttpHeaders, curie?: string): Promise<HttpResponse<PagedListResource<TListDto> | ProblemDetails>> {
    const link = this.getLinkFromHome(rel, LinkNames.getList, curie);
    const uri = link.href

    const response = await this.getListByUri<TListDto>(uri, parameters, headers);

    return response;
  }

  public async getListAsCsv(rel: string, parameters: {}, headers?: HttpHeaders, curie?: string): Promise<HttpResponse<Blob> | undefined> {
    const link = this.getLinkFromHome(rel, LinkNames.getList, curie);
    const uri = link.href;

    const response = await this.getListByUriAsCsv(uri, parameters, headers);

    return response;
  }

  public async getListByUri<TListDto extends ResourceDto>(uri: string, parameters: {}, headers?: HttpHeaders): Promise<HttpResponse<PagedListResource<TListDto> | ProblemDetails>> {
    const link = new Link();
    link.href = uri;
    const filledUri = link.fillTemplate(parameters);
    const defaultHeaders = RESTworldClient.createHeaders('application/hal+json', this._options.Version);
    const combinedHeaders = RESTworldClient.combineHeaders(headers, defaultHeaders, false);

    const response = await this.halClient.get<PagedListResource<TListDto>, ProblemDetails>(filledUri, PagedListResource, ProblemDetails, combinedHeaders);

    return response;
  }

  public async getListByUriAsCsv(uri: string, parameters: {}, headers?: HttpHeaders): Promise<HttpResponse<Blob> | undefined> {
    const link = new Link();
    link.href = uri;
    const filledUri = link.fillTemplate(parameters);
    const defaultHeaders = RESTworldClient.createHeaders('text/csv', this._options.Version);
    const combinedHeaders = RESTworldClient.combineHeaders(headers, defaultHeaders, false);

    const response = await this.halClient.httpClient.get(filledUri, { headers: combinedHeaders, responseType: 'blob', observe: 'response' }).toPromise();

    return response;
  }

  public async getAllPagesFromList<TListDto extends ResourceDto>(rel: string, parameters: {}, headers?: HttpHeaders, curie?: string): Promise<HttpResponse<PagedListResource<TListDto> | ProblemDetails>> {
    const link = this.getLinkFromHome(rel, LinkNames.getList, curie);
    const uri = link.href;

    const response = await this.getAllPagesFromListByUri<TListDto>(uri, parameters, headers);

    return response;
  }

  public async getAllPagesFromListByUri<TListDto extends ResourceDto>(uri: string, parameters: {}, headers?: HttpHeaders): Promise<HttpResponse<PagedListResource<TListDto> | ProblemDetails>> {
    const response = await this.getListByUri<TListDto>(uri, parameters, headers);

    if (!response.ok || ProblemDetails.isProblemDetails(response.body) || !response.body?._embedded?.items)
      return response;


    const items = response.body._embedded.items;
    let lastResponse = response;

    while ((lastResponse?.body?._links?.next?.length ?? 0) > 0) {
      // Get the next response
      const nextLinks = lastResponse.body?._links.next;
      const nextLink = nextLinks ? nextLinks[0]: undefined;
      const nextHref = nextLink?.href;
      if (nextHref) {
        lastResponse = await this.getListByUri<TListDto>(nextHref, parameters, headers);

        if (!lastResponse.ok || ProblemDetails.isProblemDetails(lastResponse.body) || !lastResponse.body)
          return lastResponse;

        // Combine the embedded items
        lastResponse.body?._embedded?.items?.forEach(r => items.push(r));
      }
    }

    // We combined everything, so there is just one big page
    response.body.totalPages = 1;
    response.body._links.next = undefined;

    return response;
  }

  public async getSingle(relOrUri: string, id?: number, headers?: HttpHeaders, curie?: string): Promise<HttpResponse<Resource | ProblemDetails>> {
    let uri;
    if (relOrUri.startsWith('http')) {
      if (id !== undefined)
        throw new Error('When supplying a URI, an ID cannot be supplied too.');
      if (curie)
        throw new Error('When supplying a URI, a curie cannot be supplied too.');

      uri = relOrUri;
    }
    else {
      if (!_.isNumber(id))
        throw new Error('When supplying a rel, an ID must be supplied too.');

      const link = this.getLinkFromHome(relOrUri, LinkNames.get, curie);
      uri = link.fillTemplate({ id: id.toString() });
    }

    const defaultHeaders = RESTworldClient.createHeaders('application/hal+json', this._options.Version);
    const combinedHeaders = RESTworldClient.combineHeaders(headers, defaultHeaders, false);
    const response = await this.halClient.get(uri, Resource, ProblemDetails, combinedHeaders);

    return response;
  }

  public async save(resource: Resource): Promise<HttpResponse<Resource | ProblemDetails>> {
    const saveLink = resource.findLink('save');
    if (!saveLink)
      throw new Error(`The resource ${resource} does not have a save link.`);
    if (!saveLink.name)
      throw new Error(`The save link ${saveLink} does not have a save name.`);

    const uri = saveLink.href;
    const method = saveLink.name.toLowerCase();
    const header = RESTworldClient.createHeaders('application/hal+json', this._options.Version);

    let response;
    switch (method) {
      case 'put':
        response = await this.halClient.put(uri, resource, Resource, ProblemDetails, header);
        break;
      case 'post':
        response = await this.halClient.post(uri, resource, Resource, ProblemDetails, header);
        break;
      default:
        throw new Error(`'${method}' is not allowed as link name for the save link. Only 'POST' and 'PUT' are allowed.`);
    }

    return response;
  }

  public async getAllForms(resource: Resource): Promise<HttpResponse<FormsResource | ProblemDetails>[]> {
    const urls = resource.getFormLinkHrefs();
    const header = RESTworldClient.createHeaders('application/prs.hal-forms+json', this._options.Version);
    const formsPromises = urls.map(url => this._halClient.get(url, FormsResource, ProblemDetails, header));
    const formsAndProblems = await Promise.all(formsPromises);
    return formsAndProblems;
  }

  public async submit(template: Template, formValues: {}): Promise<HttpResponse<FormsResource | ProblemDetails>> {
    const uri = template.target || '';
    const method = template.method?.toLowerCase();
    const header = RESTworldClient.createHeaders('application/prs.hal-forms+json', this._options.Version);

    let response;
    switch (method) {
      case 'put':
        response = await this.halClient.put(uri, formValues, FormsResource, ProblemDetails, header);
        break;
      case 'post':
        response = await this.halClient.post(uri, formValues, FormsResource, ProblemDetails, header);
        break;
      default:
        response = await this.halClient.get(uri, FormsResource, ProblemDetails, header);
    }

    return response;

  }

  public async delete(resource: Resource): Promise<HttpResponse<void | ProblemDetails>> {
    const deleteLink = resource.findLink('delete');
    if (!deleteLink)
      throw new Error(`The resource ${resource} does not have a delete link.`);
    const uri = deleteLink.href;
    const header = RESTworldClient.createHeaders('application/hal+json', this._options.Version);

    const response = await this.halClient.delete(uri, ProblemDetails, header);

    return response;
  }

  public getAllLinksFromHome(): { [rel: string]: Link[] | undefined; } {
    if (!this._homeResource)
      throw new Error('Home resource is not set. Call ensureHomeResourceIsSet() first.');

    return this._homeResource._links;
  }

  public getLinkFromHome(rel: string, name?: string, curie?: string): Link {
    const links = this.getLinksFromHome(rel, curie);

    const link = name ? links.find(l => l.name === name) : links[0];

    if (!link)
      throw new Error(`The home resource does not have a link with the rel '${this.getFullRel(rel, curie)}' and the name '${name}'.`);

    return link;
  }

  public getLinksFromHome(rel: string, curie?: string): Link[] {
    if (!this._homeResource)
      throw new Error('Home resource is not set. Call ensureHomeResourceIsSet() first.');

    const fullRel = this.getFullRel(rel, curie);
    const links = this._homeResource._links[fullRel];
    if (!links || links.length === 0)
      throw Error(`The home resource does not have a link with the rel '${fullRel}'.`);

    return links;
  }

  private getFullRel(rel: string, curie?: string): string {
    // rel already includes a curie => just return it
    if (rel.includes(':'))
      return rel;

    // No curie given => use default curie.
    if (!curie)
      curie = this._defaultCurie;

    // Combine curie and rel
    const fullRel = `${curie}:${rel}`;

    return fullRel;
  }

  private static createHeaders(mediaType?: 'application/hal+json' | 'application/prs.hal-forms+json' | 'text/csv', version?: number): HttpHeaders {
    if (version)
      return new HttpHeaders({
        'Accept': `${mediaType || 'application/hal+json'}; v=${version}`,
        'Content-Type': `${mediaType || 'application/hal+json'}; v=${version}`
      });
    return new HttpHeaders();
  }

  private static combineHeaders(originalHeaders: HttpHeaders | undefined, headersToAdd: HttpHeaders | undefined, overwriteExisting: boolean) {
    if (!headersToAdd)
      return originalHeaders;
    if (!originalHeaders)
      return headersToAdd;

    let combinedHeaders = originalHeaders;

    for (const key in headersToAdd.keys) {
      if (!combinedHeaders.has(key) || overwriteExisting) {
        const headerValuesToAdd = headersToAdd.getAll(key);
        if (headerValuesToAdd)
          combinedHeaders = combinedHeaders.set(key, headerValuesToAdd);
      }
    }

    return combinedHeaders;
  }
}

File: src/RESTworld/RESTworld.Client.Net/RestWorldClient.cs
using HAL.Client.Net;
using HAL.Common;
using HAL.Common.Forms;
using Microsoft.Extensions.Logging;
using RESTworld.Common.Client;
using RESTworld.Common.Dtos;
using System.Text.Json;

namespace RESTworld.Client.Net;

/// <summary>
/// A client to call a RESTworld backend.
/// </summary>
public class RestWorldClient : IRestWorldClient
{
    private readonly ApiUrl _apiUrl;
    private readonly string? _defaultCurie;
    private readonly IHalClient _halClient;
    private readonly Resource<HomeDto> _homeResource;
    private readonly ILogger<RestWorldClient> _logger;

    /// <summary>
    /// Creates a new instance of the <see cref="RestWorldClient"/> class.
    /// </summary>
    /// <param name="halClient">The underlying <see cref="IHalClient"/> to use for all calls.</param>
    /// <param name="homeResource">The home resource to construct this client for.</param>
    /// <param name="apiUrl">The URI which points to the home resource. Used for version information.</param>
    /// <param name="logger">The logger.</param>
    /// <exception cref="ArgumentNullException"></exception>
    public RestWorldClient(IHalClient halClient, Resource<HomeDto> homeResource, ApiUrl apiUrl, ILogger<RestWorldClient> logger)
    {
        _halClient = halClient ?? throw new ArgumentNullException(nameof(halClient));
        _homeResource = homeResource ?? throw new ArgumentNullException(nameof(homeResource));
        _apiUrl = apiUrl ?? throw new ArgumentNullException(nameof(apiUrl));
        _logger = logger ?? throw new ArgumentNullException(nameof(logger));

        var version = apiUrl.Version?.ToString();
        var homeVersions = homeResource.State?.Versions;
        if (version is not null && homeVersions is not null)
        {
            if (homeVersions.Deprecated is not null && homeVersions.Deprecated.Contains(version))
                logger.LogWarning("The version {Version} used for {ApiName} at {ApiUrl} is deprecated.", version, apiUrl.Name, apiUrl.Url);
            else if (homeVersions.Supported is not null && !homeVersions.Supported.Contains(version))
                logger.LogError("The version {Version} used for {ApiName} at {ApiUrl} is neither supported nor deprecated. Expect calls to fail!", version, apiUrl.Name, apiUrl.Url);
        }

        _defaultCurie = GetDefaultCurie();

        logger.LogInformation("Created a new RestWorldClient for {ApiName} at {ApiUrl} with version {Version}.", apiUrl.Name, apiUrl.Url, version);
    }

    /// <summary>
    /// Creates a new client for the given <paramref name="apiUrl"/>.
    /// </summary>
    /// <param name="halClient">The underlying <see cref="IHalClient"/> to use for all calls.</param>
    /// <param name="apiUrl">The URI which points to the home resource.</param>
    /// <param name="logger">The logger.</param>
    /// <param name="cancellationToken">The cancellation token.</param>
    /// <returns>A new client constructed from the home resource found at the given <paramref name="apiUrl"/>.</returns>
    /// <exception cref="ArgumentNullException"></exception>
    /// <exception cref="ArgumentException"></exception>
    public static async Task<RestWorldClient> CreateAsync(IHalClient halClient, ApiUrl apiUrl, ILogger<RestWorldClient> logger, CancellationToken cancellationToken = default)
    {
        ArgumentNullException.ThrowIfNull(halClient);
        ArgumentNullException.ThrowIfNull(apiUrl);
        ArgumentNullException.ThrowIfNull(logger);

        if (string.IsNullOrWhiteSpace(apiUrl.Url))
            throw new ArgumentException("The apiUrl.Url must contain a value.", nameof(apiUrl));

        var response = await halClient.GetAsync<HomeDto>(new Uri(apiUrl.Url), version: apiUrl.Version?.ToString(), cancellationToken: cancellationToken);
        if (!response.Succeeded || response.Resource is null)
        {
            var jsonOptions = new JsonSerializerOptions(JsonSerializerDefaults.Web)
            {
                WriteIndented = true
            };
            var ex = new HttpRequestException($"Unable to get the home resource from {apiUrl.Url} with version {apiUrl.Version}. Response was:{Environment.NewLine}{JsonSerializer.Serialize(response, jsonOptions)}", null, response.StatusCode);
            ex.Data["Response"] = response;
            throw ex;
        }

        return new RestWorldClient(halClient, response.Resource, apiUrl, logger);
    }

    /// <inheritdoc/>
    public Task<HalResponse<Resource<TState>>> DeleteAsync<TState>(
        Uri requestUri, string? eTag = null, IDictionary<string, object>? uriParameters = null,
        IDictionary<string, IEnumerable<string>>? headers = null, string? version = null,
        CancellationToken cancellationToken = default)
        => _halClient.DeleteAsync<TState>(requestUri, eTag, uriParameters, headers, version, cancellationToken);

    /// <inheritdoc/>
    public Task<HalResponse<Resource>> DeleteAsync(
        Uri requestUri, string? eTag = null, IDictionary<string, object>? uriParameters = null,
        IDictionary<string, IEnumerable<string>>? headers = null, string? version = null,
        CancellationToken cancellationToken = default)
        => _halClient.DeleteAsync(requestUri, eTag, uriParameters, headers, version, cancellationToken);

    /// <inheritdoc/>
    public Task<HalResponse<FormsResource<TState>>> DeleteFormAsync<TState>(
        Uri requestUri,
        string? eTag = null,
        IDictionary<string, object>? uriParameters = null,
        IDictionary<string, IEnumerable<string>>? headers = null,
        string? version = null,
        CancellationToken cancellationToken = default)
        => _halClient.DeleteFormAsync<TState>(requestUri, eTag, uriParameters, headers, version, cancellationToken);

    /// <inheritdoc/>
    public Task<HalResponse<FormsResource>> DeleteFormAsync(
        Uri requestUri,
        string? eTag = null,
        IDictionary<string, object>? uriParameters = null,
        IDictionary<string, IEnumerable<string>>? headers = null,
        string? version = null,
        CancellationToken cancellationToken = default)
        => _halClient.DeleteFormAsync(requestUri, eTag, uriParameters, headers, version, cancellationToken);

    /// <inheritdoc/>
    public IDictionary<string, ICollection<Link>> GetAllLinksFromHome() => _homeResource.Links ?? new Dictionary<string, ICollection<Link>>();

    /// <inheritdoc/>
    public async Task<HalResponse<FormsResource<Page>>> GetAllPagesFormListAsync(
        string rel,
        string? curie = null,
        IDictionary<string, object>? uriParameters = null,
        IDictionary<string, IEnumerable<string>>? headers = null,
        string? version = null,
        CancellationToken cancellationToken = default)
    {
        var response = await GetFormListAsync(rel, curie, uriParameters, headers, version, cancellationToken);

        if (!response.Succeeded || response.Resource is null)
            return response;

        if (!response.Succeeded || response.Resource is null || response.Resource.Embedded is null || !response.Resource.Embedded.TryGetValue(Common.Constants.ListItems, out var items))
            return response;

        items ??= [];

        var lastResponse = response;

        while (lastResponse.Resource!.TryFindLink("next", out var nextLink))
        {
            // Get the next response
            lastResponse = await nextLink!.FollowGetFormAsync<Page>(_halClient, null, headers, version, cancellationToken);

            if (!lastResponse.Succeeded)
                return lastResponse;

            // Combine the embedded items
            if (lastResponse.Resource?.Embedded is not null && lastResponse.Resource.Embedded.TryGetValue(Common.Constants.ListItems, out var lastItems) && lastItems is not null)
            {
                foreach (var embeddedItemResource in lastItems)
                {
                    items.Add(embeddedItemResource);
                }
            }
        }

        // We combined everything, so there is just one big page
        response.Resource.State?.TotalPages = 1;

        response.Resource.Links!.Remove("next");

        return response;
    }

    /// <inheritdoc/>
    public async Task<HalResponse<Resource<Page>>> GetAllPagesListAsync(
        string rel,
        string? curie = default,
        IDictionary<string, object>? uriParameters = default,
        IDictionary<string, IEnumerable<string>>? headers = default,
        string? version = default,
        CancellationToken cancellationToken = default)
    {
        var response = await GetListAsync(rel, curie, uriParameters, headers, version, cancellationToken);

        if (!response.Succeeded || response.Resource is null)
            return response;

        if (!response.Succeeded || response.Resource is null || response.Resource.Embedded is null || !response.Resource.Embedded.TryGetValue(Common.Constants.ListItems, out var items))
            return response;

        items ??= [];

        var lastResponse = response;

        while (lastResponse.Resource!.TryFindLink("next", out var nextLink))
        {
            // Get the next response
            lastResponse = await nextLink!.FollowGetAsync<Page>(_halClient, null, headers, version, cancellationToken);

            if (!lastResponse.Succeeded)
                return lastResponse;

            // Combine the embedded items
            if (lastResponse.Resource?.Embedded is not null && lastResponse.Resource.Embedded.TryGetValue(Common.Constants.ListItems, out var lastItems) && lastItems is not null)
            {
                foreach (var embeddedItemResource in lastItems)
                {
                    items.Add(embeddedItemResource);
                }
            }
        }

        // We combined everything, so there is just one big page
        response.Resource.State?.TotalPages = 1;

        response.Resource.Links!.Remove("next");

        return response;
    }

    /// <inheritdoc/>
    public Task<HalResponse<Resource<TState>>> GetAsync<TState>(
        Uri requestUri,
        IDictionary<string, object>? uriParameters = null,
        IDictionary<string, IEnumerable<string>>? headers = null,
        string? version = null,
        CancellationToken cancellationToken = default)
        => _halClient.GetAsync<TState>(requestUri, uriParameters, headers, version, cancellationToken);

    /// <inheritdoc/>
    public Task<HalResponse<Resource>> GetAsync(
        Uri requestUri,
        IDictionary<string, object>? uriParameters = null,
        IDictionary<string, IEnumerable<string>>? headers = null,
        string? version = null,
        CancellationToken cancellationToken = default)
        => _halClient.GetAsync(requestUri, uriParameters, headers, version, cancellationToken);

    /// <inheritdoc/>
    public Task<HalResponse<FormsResource<TState>>> GetFormAsync<TState>(
        Uri requestUri,
        IDictionary<string, object>? uriParameters = null,
        IDictionary<string, IEnumerable<string>>? headers = null,
        string? version = null,
        CancellationToken cancellationToken = default)
        => _halClient.GetFormAsync<TState>(requestUri, uriParameters, headers, version, cancellationToken);

    /// <inheritdoc/>
    public Task<HalResponse<FormsResource>> GetFormAsync(
        Uri requestUri,
        IDictionary<string, object>? uriParameters = null,
        IDictionary<string, IEnumerable<string>>? headers = null,
        string? version = null,
        CancellationToken cancellationToken = default)
        => _halClient.GetFormAsync(requestUri, uriParameters, headers, version, cancellationToken);

    /// <inheritdoc/>
    public async Task<HalResponse<FormsResource<Page>>> GetFormListAsync(
        string rel,
        string? curie = null,
        IDictionary<string, object>? uriParameters = null,
        IDictionary<string, IEnumerable<string>>? headers = null,
        string? version = null,
        CancellationToken cancellationToken = default)
    {
        var link = GetLinkFromHome(rel, Common.Constants.GetListLinkName, curie)
            ?? throw new ArgumentException($"Unable to find a link with the rel {rel} and the curie {curie ?? "null"}", nameof(rel));
        var response = await link.FollowGetFormAsync<Page>(_halClient, uriParameters, headers, version, cancellationToken);

        return response;
    }

    /// <inheritdoc/>
    public Link? GetLinkFromHome(string rel, string? name = default, string? curie = default)
    {
        var fullRel = GetFullRel(rel, curie);
        var link = _homeResource.FindLink(fullRel, name);

        return link;
    }

    /// <inheritdoc/>
    public IEnumerable<Link> GetLinksFromHome(string rel, string? curie = default)
    {
        var fullRel = GetFullRel(rel, curie);
        var links = _homeResource.FindLinks(fullRel);

        return links;
    }

    /// <inheritdoc/>
    public async Task<HalResponse<Resource<Page>>> GetListAsync(
        string rel,
        string? curie = default,
        IDictionary<string, object>? uriParameters = default,
        IDictionary<string, IEnumerable<string>>? headers = default,
        string? version = default,
        CancellationToken cancellationToken = default)
    {
        var link = GetLinkFromHome(rel, Common.Constants.GetListLinkName, curie)
            ?? throw new ArgumentException($"Unable to find a link with the rel {rel} and the curie {curie ?? "null"}", nameof(rel));
        var response = await link.FollowGetAsync<Page>(_halClient, uriParameters, headers, version, cancellationToken);

        return response;
    }

    /// <inheritdoc/>
    public Task<HalResponse<Resource<TState>>> PostAsync<TRequest, TState>(
        Uri requestUri, TRequest? content = default, IDictionary<string, object>? uriParameters = null,
        IDictionary<string, IEnumerable<string>>? headers = null, string? version = null,
        CancellationToken cancellationToken = default)
        => _halClient.PostAsync<TRequest, TState>(requestUri, content, uriParameters, headers, version, cancellationToken);

    /// <inheritdoc/>
    public Task<HalResponse<Resource>> PostAsync<TRequest>(
        Uri requestUri, TRequest? content = default, IDictionary<string, object>? uriParameters = null,
        IDictionary<string, IEnumerable<string>>? headers = null, string? version = null,
        CancellationToken cancellationToken = default)
        => _halClient.PostAsync(requestUri, content, uriParameters, headers, version, cancellationToken);

    /// <inheritdoc/>
    public Task<HalResponse<FormsResource<TState>>> PostFormAsync<TRequest, TState>(
        Uri requestUri,
        TRequest? content = default,
        IDictionary<string, object>? uriParameters = null,
        IDictionary<string, IEnumerable<string>>? headers = null,
        string? version = null,
        CancellationToken cancellationToken = default)
     => _halClient.PostFormAsync<TRequest, TState>(requestUri, content, uriParameters, headers, version, cancellationToken);

    /// <inheritdoc/>
    public Task<HalResponse<FormsResource>> PostFormAsync<TRequest>(
        Uri requestUri,
        TRequest? content = default,
        IDictionary<string, object>? uriParameters = null,
        IDictionary<string, IEnumerable<string>>? headers = null,
        string? version = null,
        CancellationToken cancellationToken = default)
        => _halClient.PostFormAsync(requestUri, content, uriParameters, headers, version, cancellationToken);

    /// <inheritdoc/>
    public Task<HalResponse<Resource<TState>>> PutAsync<TRequest, TState>(
        Uri requestUri, TRequest? content = default, IDictionary<string, object>? uriParameters = null,
        IDictionary<string, IEnumerable<string>>? headers = null, string? version = null,
        CancellationToken cancellationToken = default)
        => _halClient.PutAsync<TRequest, TState>(requestUri, content, uriParameters, headers, version, cancellationToken);

    /// <inheritdoc/>
    public Task<HalResponse<Resource>> PutAsync<TRequest>(
        Uri requestUri, TRequest? content = default, IDictionary<string, object>? uriParameters = null,
        IDictionary<string, IEnumerable<string>>? headers = null, string? version = null,
        CancellationToken cancellationToken = default)
        => _halClient.PutAsync(requestUri, content, uriParameters, headers, version, cancellationToken);

    /// <inheritdoc/>
    public Task<HalResponse<FormsResource<TState>>> PutFormAsync<TRequest, TState>(
        Uri requestUri,
        TRequest? content = default,
        IDictionary<string, object>? uriParameters = null,
        IDictionary<string, IEnumerable<string>>? headers = null,
        string? version = null,
        CancellationToken cancellationToken = default)
        => _halClient.PutFormAsync<TRequest, TState>(requestUri, content, uriParameters, headers, version, cancellationToken);

    /// <inheritdoc/>
    public Task<HalResponse<FormsResource>> PutFormAsync<TRequest>(
        Uri requestUri,
        TRequest? content = default,
        IDictionary<string, object>? uriParameters = null,
        IDictionary<string, IEnumerable<string>>? headers = null,
        string? version = null,
        CancellationToken cancellationToken = default)
        => _halClient.PutFormAsync(requestUri, content, uriParameters, headers, version, cancellationToken);

    /// <inheritdoc/>
    public Task<HalResponse<Resource<TState>>> SendAsync<TRequest, TState>(
        HttpMethod method,
        Uri requestUri,
        TRequest? content = default,
        IDictionary<string, object>? uriParameters = null,
        IDictionary<string, IEnumerable<string>>? headers = null,
        string? version = null,
        CancellationToken cancellationToken = default)
        => _halClient.SendAsync<TRequest, TState>(method, requestUri, content, uriParameters, headers, version, cancellationToken);

    /// <inheritdoc/>
    public Task<HalResponse<Resource>> SendAsync<TRequest>(
        HttpMethod method,
        Uri requestUri,
        TRequest? content = default,
        IDictionary<string, object>? uriParameters = null,
        IDictionary<string, IEnumerable<string>>? headers = null,
        string? version = null,
        CancellationToken cancellationToken = default)
        => _halClient.SendAsync(method, requestUri, content, uriParameters, headers, version, cancellationToken);

    /// <inheritdoc/>
    public Task<HalResponse<FormsResource<TState>>> SendFormAsync<TRequest, TState>(
        HttpMethod method,
        Uri requestUri,
        TRequest? content = default,
        IDictionary<string, object>? uriParameters = null,
        IDictionary<string, IEnumerable<string>>? headers = null,
        string? version = null,
        CancellationToken cancellationToken = default)
        => _halClient.SendFormAsync<TRequest, TState>(method, requestUri, content, uriParameters, headers, version, cancellationToken);

    /// <inheritdoc/>
    public Task<HalResponse<FormsResource>> SendFormAsync<TRequest>(
        HttpMethod method,
        Uri requestUri,
        TRequest? content = default,
        IDictionary<string, object>? uriParameters = null,
        IDictionary<string, IEnumerable<string>>? headers = null,
        string? version = null,
        CancellationToken cancellationToken = default)
        => _halClient.SendFormAsync(method, requestUri, content, uriParameters, headers, version, cancellationToken);

    /// <inheritdoc/>
    public Task<HttpResponseMessage> SendHttpRequestAsync<TRequest>(
        HttpMethod method,
        Uri requestUri,
        Accepts accepts,
        TRequest? content = default,
        IDictionary<string, object>? uriParameters = default,
        IDictionary<string, IEnumerable<string>>? headers = default,
        string? version = default,
        HttpCompletionOption completionOption = HttpCompletionOption.ResponseContentRead,
        CancellationToken cancellationToken = default)
        => _halClient.SendHttpRequestAsync(method, requestUri, accepts, content, uriParameters, headers, version, completionOption, cancellationToken);

    /// <inheritdoc/>
    public Task<HttpResponseMessage> SendHttpRequestAsync<TRequest>(
        HttpRequestMessage request,
        Accepts accepts,
        TRequest? content = default,
        IDictionary<string, object>? uriParameters = null,
        IDictionary<string, IEnumerable<string>>? headers = null,
        string? version = null,
        HttpCompletionOption completionOption = HttpCompletionOption.ResponseContentRead,
        CancellationToken cancellationToken = default)
        => _halClient.SendHttpRequestAsync(request, accepts, content, uriParameters, headers, version, completionOption, cancellationToken);

    private string? GetDefaultCurie()
    {
        if (_homeResource is not null && _homeResource.TryFindLink("curies", out var link))
            return link?.Name;

        _logger.LogWarning("Unable to determine the default curie for the home resource at {Url} with version {Version}, because the home resource does not have a curies link.", _apiUrl.Url, _apiUrl.Version);
        return null;
    }

    private string GetFullRel(string rel, string? curie)
    {
        // rel already includes a curie => just return it
        if (rel.Contains(':'))
            return rel;

        // No curie given => use default curie.
        curie ??= _defaultCurie;

        // Combine curie and rel
        var fullRel = string.Concat(curie, ":", rel);

        return fullRel;
    }
}